firedrake/cython/dmcommon.pyx

```diff
@@ -666,6 +668,174 @@ def quadrilateral_closure_ordering(PETSc.DM plex,
     return cell_closure
 
 
+cdef inline PetscInt _get_simplex_orientation(np.ndarray[PetscInt, ndim=2, mode="c"] cell_closure,
+                                              PetscInt cell,
+                                              PetscInt eStart,
+                                              PetscInt eEnd,
+                                              PetscInt *cone,
+                                              PetscInt coneSize):
+    cdef:
+        PetscInt k, n, e, q, o = 0, coneSize1 = coneSize
+        PetscInt *cone1 = NULL
+        PetscInt *inds = NULL
+
+    CHKERR(PetscMalloc1(coneSize, &cone1))
+    CHKERR(PetscMalloc1(coneSize, &inds))
+    for k in range(coneSize1):
+        cone1[k] = cone[k]
+    n = 0
+    for e in range(eStart, eEnd):
+        q = cell_closure[cell, e]
+        for k in range(coneSize1):
+            if q == cone1[k]:
+                inds[n] = k
+                n += 1
+                break
+        else:
+            continue
+        while k < coneSize1 - 1:
+            cone1[k] = cone1[k + 1]
+            k += 1
+        coneSize1 -= 1
+    assert n == coneSize
+    for k in range(n):
+        o += np.math.factorial(n - 1 - k) * inds[k]
+    CHKERR(PetscFree(cone1))
+    CHKERR(PetscFree(inds))
+    return o
+
+
+cdef inline PetscInt _ncr(PetscInt n , PetscInt r):
+    return np.math.factorial(n) // np.math.factorial(r) // np.math.factorial(n - r)
+
+
+@cython.boundscheck(False)
+@cython.wraparound(False)
+@cython.cdivision(True)
+def entity_orientations(mesh,
+                        np.ndarray[PetscInt, ndim=2, mode="c"] cell_closure):
+    """Compute entity orientations.
+
+    :arg mesh: The `MeshTopology` object encapsulating the mesh topology
+    :arg cell_closure: The two-dimensional array, each row of which contains
+        the closure of the associated cell
+
+    Each row of cell_closure must have been ordered to work well with
+    FInAT entity_dofs to create cell_node map (see `get_cell_nodes`).
```

Comment:
I looked at that, but didn't see much that was enlightening.

Can you describe the return value a bit more?

Reply:
So I've now added more words at `AbstractMeshTopology.entity_permutations`, so here I just added `See :meth:~.AbstractMeshTopology.entity_permutations`